File: src/test/kotlin/com/gradlewatcher/domain/RedactorTest.kt
package com.gradlewatcher.domain

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

class RedactorTest {

    @Test
    fun `masks -P and -D secret values, leaves safe flags`() {
        val cmd = "gradlew publish -Psigning.password=abc -Dtoken=xyz -PsafeFlag=ok --info"
        val out = Redactor.redactCommandLine(cmd)
        assertTrue(out.contains("-Psigning.password=***"), out)
        assertTrue(out.contains("-Dtoken=***"), out)
        assertTrue(out.contains("-PsafeFlag=ok"), out)
        assertTrue(out.contains("--info"), out)
        assertTrue(!out.contains("abc") && !out.contains("xyz"), out)
    }

    @Test
    fun `masks long-option --key=value secrets`() {
        val out = Redactor.redactToken("--repository-password=hunter2")
        assertEquals("--repository-password=***", out)
    }

    @Test
    fun `masks credentials embedded in URLs`() {
        val out = Redactor.redactToken("https://user:pw@repo.example.com/path")
        assertEquals("https://***:***@repo.example.com/path", out)
    }

    @Test
    fun `leaves non-sensitive tokens unchanged`() {
        assertEquals("clean", Redactor.redactToken("clean"))
        assertEquals("-Dfile.encoding=UTF-8", Redactor.redactToken("-Dfile.encoding=UTF-8"))
    }

    @Test
    fun `masks tab-separated secret flags, not just space-separated`() {
        val out = Redactor.redactCommandLine("java\t-Dapi.token=SECRET\t-jar app.jar")
        assertTrue(out.contains("-Dapi.token=***"), out)
        assertTrue(!out.contains("SECRET"), out)
    }

    @Test
    fun `does not over-redact benign keys containing the substring key`() {
        assertEquals("-Dmonkey.count=5", Redactor.redactToken("-Dmonkey.count=5"))
        assertEquals("-Dkeystore.path=/x", Redactor.redactToken("-Dkeystore.path=/x"))
        // a real key-named property is still masked
        assertEquals("-Psigning.key=***", Redactor.redactToken("-Psigning.key=abc"))
    }

    @Test
    fun `redacts secrets inside a log line`() {
        val line = "2026-06-23T10:25:24 [INFO] [Build] running with -Papikey=deadbeef now"
        val out = Redactor.redactLogLine(line)
        assertTrue(out.contains("-Papikey=***"), out)
        assertTrue(!out.contains("deadbeef"), out)
    }
}

File: spikes/go-core/internal/poll/redactor_test.go
package poll_test

import (
	"strings"
	"testing"

	"github.com/cdsap/daemonitor/spikes/go-core/internal/poll"
)

// Fixtures mirror core/.../RedactorTest.kt
func TestRedactCommandLineParity(t *testing.T) {
	cmd := "gradlew publish -Psigning.password=abc -Dtoken=xyz -PsafeFlag=ok --info"
	out := poll.RedactCommandLine(cmd)
	if !strings.Contains(out, "-Psigning.password=***") {
		t.Fatalf("missing password mask: %s", out)
	}
	if !strings.Contains(out, "-Dtoken=***") {
		t.Fatalf("missing token mask: %s", out)
	}
	if !strings.Contains(out, "-PsafeFlag=ok") {
		t.Fatalf("safe flag changed: %s", out)
	}
	if !strings.Contains(out, "--info") {
		t.Fatalf("info flag missing: %s", out)
	}
	if strings.Contains(out, "abc") || strings.Contains(out, "xyz") {
		t.Fatalf("secret leaked: %s", out)
	}
}

func TestRedactTokenLongOption(t *testing.T) {
	got := poll.RedactToken("--repository-password=hunter2")
	if got != "--repository-password=***" {
		t.Fatalf("got %q", got)
	}
}

func TestRedactTokenURLCredentials(t *testing.T) {
	got := poll.RedactToken("https://user:pw@repo.example.com/path")
	if got != "https://***:***@repo.example.com/path" {
		t.Fatalf("got %q", got)
	}
}

func TestRedactTokenLeavesSafe(t *testing.T) {
	if poll.RedactToken("clean") != "clean" {
		t.Fatal("clean changed")
	}
	if poll.RedactToken("-Dfile.encoding=UTF-8") != "-Dfile.encoding=UTF-8" {
		t.Fatal("encoding changed")
	}
}

func TestRedactCommandLineTabs(t *testing.T) {
	out := poll.RedactCommandLine("java\t-Dapi.token=SECRET\t-jar app.jar")
	if !strings.Contains(out, "-Dapi.token=***") {
		t.Fatalf("missing mask: %s", out)
	}
	if strings.Contains(out, "SECRET") {
		t.Fatalf("secret leaked: %s", out)
	}
}

func TestRedactDoesNotOverRedactKeySubstring(t *testing.T) {
	if poll.RedactToken("-Dmonkey.count=5") != "-Dmonkey.count=5" {
		t.Fatal("monkey.count over-redacted")
	}
	if poll.RedactToken("-Dkeystore.path=/x") != "-Dkeystore.path=/x" {
		t.Fatal("keystore.path over-redacted")
	}
	if poll.RedactToken("-Psigning.key=abc") != "-Psigning.key=***" {
		t.Fatal("signing.key not masked")
	}
}

func TestRedactLogLine(t *testing.T) {
	line := "2026-06-23T10:25:24 [INFO] [Build] running with -Papikey=deadbeef now"
	out := poll.RedactLogLine(line)
	if !strings.Contains(out, "-Papikey=***") {
		t.Fatalf("missing mask: %s", out)
	}
	if strings.Contains(out, "deadbeef") {
		t.Fatalf("secret leaked: %s", out)
	}
}
